src/lib/protocols/quic.c

```diff
@@ -1045,27 +1045,26 @@ static int __reassemble(struct ndpi_flow_struct *flow, const u_int8_t *frag,
     if(!flow->l4.udp.quic_reasm_buf)
       return -1; /* Memory error */
     flow->l4.udp.quic_reasm_buf_len = 0;
+    flow->l4.udp.quic_reasm_buf_last_pos = 0;
   }
-  if(flow->l4.udp.quic_reasm_buf_len != frag_offset)
-    return -2; /* Out-of-order, retransmission, overlapping */
-  if(frag_offset + frag_len > max_quic_reasm_buffer_len)
+  if(last_pos > max_quic_reasm_buffer_len)
     return -3; /* Buffer too small */
 
-  memcpy(&flow->l4.udp.quic_reasm_buf[flow->l4.udp.quic_reasm_buf_len],
-	 frag, frag_len);
+  memcpy(&flow->l4.udp.quic_reasm_buf[frag_offset], frag, frag_len);
   flow->l4.udp.quic_reasm_buf_len += frag_len;
+  flow->l4.udp.quic_reasm_buf_last_pos = last_pos > flow->l4.udp.quic_reasm_buf_last_pos ? last_pos : flow->l4.udp.quic_reasm_buf_last_pos;
 
   *buf = flow->l4.udp.quic_reasm_buf;
   *buf_len = flow->l4.udp.quic_reasm_buf_len;
   return 0;
 }
-static int is_ch_complete(const u_int8_t *buf, uint64_t buf_len)
+static int is_ch_complete(const u_int8_t *buf, uint64_t buf_len, uint64_t last_pos)
 {
   uint32_t msg_len;
 
   if(buf_len >= 4) {
     msg_len = (buf[1] << 16) + (buf[2] << 8) + buf[3];
-    if (4 + msg_len == buf_len) {
+    if (last_pos == buf_len &&  4 + msg_len == buf_len) {
```

Comment:
You need to move this new check in the previous "if" above; otherwise we end up reading un-initialized buffer (as you can easily check with valgrind)

Reply:
done!